src/main/scala/firrtl/passes/memlib/ReplaceMemMacros.scala

```diff
@@ -242,18 +248,24 @@ class ReplaceMemMacros(writer: ConfWriter) extends Transform with DependencyAPIM
           val newWrapperName = nameMap(mname -> m.name)
           val newMemBBName = namespace.newName(s"${newWrapperName}_ext")
           val newMem = m.copy(name = newMemBBName)
-          memMods ++= createMemModule(newMem, newWrapperName)
+          memMods ++= createMemModule(newMem, newWrapperName, insertFunc)
           WDefInstance(m.info, m.name, newWrapperName, UnknownType)
         case Some((module, mem)) =>
           WDefInstance(m.info, m.name, nameMap(module -> mem), UnknownType)
       }
-    case sx => sx.map(updateMemStmts(namespace, nameMap, mname, memPortMap, memMods))
+    case sx => sx.map(updateMemStmts(namespace, nameMap, mname, memPortMap, memMods, insertFunc))
   }
 
-  def updateMemMods(namespace: Namespace, nameMap: NameMap, memMods: Modules)(m: DefModule) = {
+  def updateMemMods(
+    namespace:  Namespace,
+    nameMap:    NameMap,
+    memMods:    Modules,
+    insertFunc: DefAnnotatedMemory => Unit
+  )(m:          DefModule
+  ) = {
```

Comment:
I think these methods should be private (since we're changing them anyway, might as well take the opportunity).

Also, rather than using `insertFunc` to mutate an Annotation, why not just pass a `ListBuffer[DefAnnotatedMemory]` here and get rid of `AnnotatedMemoriesCollectorAnnotation` altogether? It's all private, I don't think there's any need for that annotation when we could just use the mutable collection directly and construct `MemLibOutConfigFileAnnotation` at the end.

Reply:
> get rid of `AnnotatedMemoriesCollectorAnnotation`

I personal prefer using an standalone Annotation to store this mutable data,so in the `DumpMemoryAnnotations` Transform, we can directly convert this annotation to other annotations, in current case only the configure file, in the next PRs, `InlinedBlackBoxes` can be converted to.
That’s the reason why I add this intermediate annotation.

Reply:
And another question is if we use mutable collection here, when I switch to dependency API, will it be thread safe? As @seldridge said, 
> I think there can be bugs when working in a parallel environment using the same stage. This can happen in tests.

this is the reason I put mutable thing in annotations, which makes transform fully immutable.

Reply:
> I think these methods should be private (since we're changing them anyway, might as well take the opportunity).

Sure, I’ll add private keywords when finishing review for a easy rebasing flow.

Reply:
> And another question is if we use mutable collection here, when I switch to dependency API, will it be thread safe?

You just need to create one instance of the collection for every call to the `execute` function. Like do not make it a member of the class, instead create it in `execute` and give a reference to it to every function that needs to mutate.

Reply:
Got it, fixed :)

Reply:
Hi, @jackkoenig 
> Also, rather than using insertFunc to mutate an Annotation, why not just pass a ListBuffer[DefAnnotatedMemory] here and get rid of AnnotatedMemoriesCollectorAnnotation altogether? 

Done. I have removed that annotation. But still have the `AnnotatedMemoriesAnnotation` For transform to other annotations. Is that ok? As @ekiwi suggested, mutable is in execute function now. 

> It's all private, I don't think there's any need for that annotation when we could just use the mutable collection directly and construct MemLibOutConfigFileAnnotation at the end.

Since after this PR, I wanna transform to Verilog blackboxes annotations, thus create the intermediate annotation, and consume it in next transform.